apps/server/src/shared/domain/service/file-copy-append.service.ts

```diff
@@ -72,23 +76,151 @@ export class FileCopyAppendService {
 			elements,
 		};
 		taskCopyStatus.status = this.copyHelperService.deriveStatusFromElements(taskCopyStatus.elements as CopyStatus[]);
-		taskCopyStatus.elements = this.replaceFileGroup(taskCopyStatus.elements, updatedFileGroupStatus);
+		taskCopyStatus.elements = this.setFileGroupStatus(taskCopyStatus.elements, updatedFileGroupStatus);
 		return taskCopyStatus;
 	}
 
-	private getFileGroupStatus(elements: CopyStatus[] = []): CopyStatus {
-		return elements?.find((el) => el.type === CopyElementType.FILE_GROUP) as CopyStatus;
+	private getFileGroupStatus(elements: CopyStatus[] = []): CopyStatus | undefined {
+		return elements.find((el) => el.type === CopyElementType.FILE_GROUP) as CopyStatus;
 	}
 
-	private replaceFileGroup(elements: CopyStatus[] = [], fileGroupStatus: CopyStatus): CopyStatus[] {
-		return elements.map((el) => (el.type === CopyElementType.FILE_GROUP ? fileGroupStatus : el));
+	private setFileGroupStatus(elements: CopyStatus[] = [], updatedFileGroupStatus: CopyStatus): CopyStatus[] {
+		const fileGroupStatus = this.getFileGroupStatus(elements);
+		if (fileGroupStatus) {
+			return elements.map((el) => (el.type === CopyElementType.FILE_GROUP ? updatedFileGroupStatus : el));
+		}
+		return [...elements, updatedFileGroupStatus];
 	}
 
 	private createFileStatuses(files: FileDto[]): CopyStatus[] {
 		return files.map((file) => ({
 			type: CopyElementType.FILE,
-			name: file.name,
+			title: file.name,
 			status: CopyStatusEnum.SUCCESS,
 		}));
 	}
+
+	private createFileStatusesByCopyResult(files: FileLegacyResponse[]): CopyStatus[] {
+		return files.map(({ oldFileId, fileId, filename }) => ({
+			type: CopyElementType.FILE,
+			status: fileId ? CopyStatusEnum.SUCCESS : CopyStatusEnum.FAIL,
+			title: filename ?? `(old fileid: ${oldFileId})`,
+		}));
+	}
+
+	async copyFiles(copyStatus: CopyStatus, courseId: EntityId, userId: EntityId, jwt: string): Promise<CopyStatus> {
+		if (copyStatus.type === CopyElementType.LESSON) {
+			return this.copyEmbeddedFilesOfLessons(copyStatus, courseId, userId);
+		}
+		if (copyStatus.type === CopyElementType.TASK) {
+			const updatedStatus = await this.appendFilesToTask(copyStatus, jwt);
+			return this.copyEmbeddedFilesOfTasks(updatedStatus, courseId, userId);
+		}
+		if (copyStatus.elements && copyStatus.elements.length > 0) {
+			copyStatus.elements = await Promise.all(
+				copyStatus.elements.map((el) => this.copyFiles(el, courseId, userId, jwt))
+			);
+			copyStatus.status = this.copyHelperService.deriveStatusFromElements(copyStatus.elements);
+		}
+		return Promise.resolve(copyStatus);
+	}
+
+	async copyEmbeddedFilesOfLessons(
+		lessonCopyStatus: CopyStatus,
+		courseId: EntityId,
+		userId: EntityId
+	): Promise<CopyStatus> {
+		const lesson = lessonCopyStatus.copyEntity as Lesson;
+		const legacyFileIds: string[] = [];
+
+		lesson.contents.forEach((item: IComponentProperties) => {
```

Comment:
This part include different abstraction level. 
copyEmbeddedFilesOfLessons should only include orchestration code like named function with do this then do this and so on. The part of each loop and each if check with more code inside can be moved to additional named functions. 

This improve the readability and the mainability of the code. (and make it easer to review :P )

Reply:
I totally agree. Especially as copyEmbeddedFilesOfLessons and copyEmbeddedFilesOfTasks share much of their structure. Already tried to refactor that on wednesday, but removed the changes as I was not able to finish it in time.